### DISCUSSION TITLE
Upgrade from 3.4 to 4.* without loosing all settings possible ?

### DISCUSSION THREAD

**Discussion by patrick-foubert-56770** [Blackymas/NSPanel_HA_Blueprint#1100] [General]
Hi, 2 weeks ago i made the setup on version 3.4 for first device, is it possible to migrate blueprint to 4.x without loosing all settings from automation ?
I made a copy for yaml automation, i presume if i want upgrade i need delete automation, blueprint and reimport from latest version (ok for esp and tft update)

**Comment by edwardtfn** [comment]
This is a nice topic. 😃 

In short, you shouldn't loose any of your settings when updating, but I will explain a little bit each one of the components, so you can understand what to expect.

### TFT
The TFT files doesn't retain any settings, so this question kind of didn't apply for TFT files.
However, there is a protocol between ESPHome and the Nextion display (where the TFT files are executed) and this is based on components set on the TFT. We can change some of those components in an update, so if you have created custom calls to the screen, you might have to review.
As this is quite low level in this project, we don't use to report the changes on the TFT on the breaking changes.

### ESPHome
The ESPHome settings consists in 2 parts, one containing your settings and a reference to the second one, the remote packages. An update are typically done on the remote package only, so your settings should always be preserved.
ESPHome publishes services and entities to Home Assistant and tries to remember your previous selections, so if you have previously enables notification sounds, as an example, that selection should be remembered after the update.
We do changes on some entities and services when upgrading, and the more high level changes are all listed on the "Breaking Changes" part of the release notes. From v3.4 to v4.0 we have plenty of those changes, but unless you have external (custom made) automations using those, you will only notice a couple of those changes, and in this case I will highlight the change on Sleep Mode, which was set as a switch on v3.4 (so it could only be `on` or `off`, while in v4.0 it is a selection of number of seconds for sleep (where 0s means "never sleeps"), so you will have to select your preference again in this case and, if using the older entity in some automation, you will have to adapt.

### Blueprint
The blueprint also consist in 2 parts, one is the blueprint file itself, the one that you have to replace to the latest version and which doesn't retain any settings, and the other part is an entry in your `automations.yaml` file with a reference to the blueprint and all of your inputs. This second part is preserved unless you delete the automation (which is not necessary), so your settings would be preserved.
Again, we might change the format of some inputs in an update, then you will have to apply your settings for that specific input again. There's no case like this in the top of my memory about migrating from v3.4 to v4.0, so you should be good.


&nbsp;
We always try to make those upgrades as smooth as possible, in a way you preserve your settings and can use the system right way. Installing a new TFT in ESPHome is enough pain for an upgrade, so we don't wanna make it worse. 😉 
When we break something important, we inform about that in the "Breaking changes" in the release notes.

**Comment by patrick-foubert-56770** [comment]
Thanks for quick and precise reply, it's logical for me, need only to update blueprint, configuration is stored on automation :) But when y try to delete and replace blueprint i got an error "Error handling message: Blueprint in use (unknown_error)" probably because any automation is linked to it. Tried also by disable automation. I take an backup of automation.yaml and try by this method and replace values on new automation with old for correponding entities

**Reply by edwardtfn** [reply]
You can just replace the files: https://github.com/Blackymas/NSPanel_HA_Blueprint/wiki/(EN)-HowTo---All-important-thing-you-should-know#2-update-blueprint

**Comment by patrick-foubert-56770** [comment]
Thanks :) I made backup and follow it !

**Reply by Doughboy68** [reply]
I updated three panels from 3.4 to 4.0 with no issues.
I just went in and cleaned up all the unused entities after.
Have to adjust automations for putting panel to sleep because new method, but other than that it was smooth.
